File: App/mobile/lib/data/models/event.dart
import 'package:json_annotation/json_annotation.dart';

part 'event.g.dart';

@JsonSerializable()
class Event {
  final String id;
  @JsonKey(name: 'creator_id')
  final String? creatorId;
  final String title;
  final String? description;
  @JsonKey(name: 'image_url')
  final String? imageUrl;
  @JsonKey(name: 'is_free')
  final bool isFree;
  @JsonKey(name: 'start_date')
  final DateTime startDate;
  @JsonKey(name: 'end_date')
  final DateTime? endDate;
  @JsonKey(name: 'location_name')
  final String? locationName;
  final String category;
  @JsonKey(name: 'event_url')
  final String? eventUrl;
  final String status;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const Event({
    required this.id,
    this.creatorId,
    required this.title,
    this.description,
    this.imageUrl,
    required this.isFree,
    required this.startDate,
    this.endDate,
    this.locationName,
    required this.category,
    this.eventUrl,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Event.fromJson(Map<String, dynamic> json) => _$EventFromJson(json);
  Map<String, dynamic> toJson() => _$EventToJson(this);
}

@JsonSerializable()
class EventListResponse {
  final List<Event> events;
  final int total;
  final int offset;
  final int limit;

  const EventListResponse({
    required this.events,
    required this.total,
    required this.offset,
    required this.limit,
  });

  factory EventListResponse.fromJson(Map<String, dynamic> json) =>
      _$EventListResponseFromJson(json);
}

File: App/backend/schemas/event.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from uuid import UUID


class EventCreate(BaseModel):
    title: str = Field(..., min_length=3, max_length=255)
    description: Optional[str] = None
    category: str = Field(..., max_length=100)
    start_date: datetime
    end_date: Optional[datetime] = None
    location_name: Optional[str] = None
    lat: Optional[float] = Field(None, ge=-90, le=90)
    lng: Optional[float] = Field(None, ge=-180, le=180)
    image_url: Optional[str] = None
    is_free: bool = True
    event_url: Optional[str] = None


class EventUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=3, max_length=255)
    description: Optional[str] = None
    category: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    location_name: Optional[str] = None
    lat: Optional[float] = Field(None, ge=-90, le=90)
    lng: Optional[float] = Field(None, ge=-180, le=180)
    image_url: Optional[str] = None
    is_free: Optional[bool] = None


class EventResponse(BaseModel):
    id: UUID
    creator_id: Optional[UUID] = None
    title: str
    description: Optional[str] = None
    image_url: Optional[str] = None
    is_free: bool
    start_date: datetime
    end_date: Optional[datetime] = None
    location_name: Optional[str] = None
    category: str
    event_url: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EventListResponse(BaseModel):
    events: List[EventResponse]
    total: int
    offset: int
    limit: int


class RSVPRequest(BaseModel):
    status: str = Field(..., pattern=r"^(attending|interested|declined)$")


class RSVPResponse(BaseModel):
    id: UUID
    user_id: UUID
    event_id: UUID
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class InteractionLogRequest(BaseModel):
    action: str = Field(..., pattern=r"^(view|click|dismiss)$")
